datafusion/functions/src/unicode/substr.rs

```diff
@@ -197,6 +213,29 @@ fn string_view_substr(
 
     let start_array = as_int64_array(&args[0])?;
 
+    // Notes for ASCII-only optimization:
+    //
+    // String characters are variable length encoded in UTF-8, `substr()` function's
+    // arguments are character-based, converting them into byte-based indices
+    // requires expensive decoding.
+    // However, checking if a string is ASCII-only is relatively cheap.
+    // If strings are ASCII only, use byte-based indices instead.
+    //
+    // A common pattern to call `substr()` is taking a small prefix of a long
+    // string, such as `substr(long_str_with_1k_chars, 1, 32)`.
+    // In such case the overhead of ASCII-validation may not be worth it, so
+    // skip the validation for long strings for now.
+    // TODO: A better heuristic is to use the ratio to decide whether to validate
+    // like `(start + count) / estimate_avg_strlen > threshold`, but it requires
+    // specialized implementation for `ScalarValue` input.
+    let estimate_avg_strlen =
```

Comment:
Can we please pull this check into its own method (so it is easier to find), perhaps something like

```rust
fn enable_ascii_fast_path(arr: &StringViewArray) -> bool { 
...
}
```